stan/math/rev/mat/functor/adj_jac_apply.hpp

```diff
@@ -230,48 +333,107 @@ struct adj_jac_vari : public vari {
     prepare_x_vis(args...);
   }
 
+  /**
+   * The constructor initializes is_var_ with whether or not the scalar type in
+   * each argument is a var
+   */
+  adj_jac_vari()
+      : vari(std::numeric_limits<double>::quiet_NaN()),  // The val_ in this
+                                                         // vari is unused
+        is_var_({{is_var<typename scalar_type<Targs>::type>::value...}}),
+        x_vis_(NULL),
+        y_vi_(NULL) {}
+
+  /**
+   * build_return_varis_and_vars returns a var with a new vari that has the
+   * value val_y
+   *
+   * @param val_y output of F::operator()
+   * @return var
+   */
+  var build_return_varis_and_vars(const double& val_y) {
+    y_vi_ = ChainableStack::instance().memalloc_.alloc_array<vari*>(1);
+    y_vi_[0] = new vari(val_y, false);
+
+    return y_vi_[0];
+  }
+
+  /**
+   * build_return_varis_and_vars constructs a container of vars of the same
+   * shape and size as the input, allocates new varis for these vars, and
+   * assigns them the values of val_y
+   *
+   * @param val_y output of F::operator()
+   * @return std::vector of vars
+   */
+  std::vector<var> build_return_varis_and_vars(
+      const std::vector<double>& val_y) {
+    M_[0] = val_y.size();
+    std::vector<var> var_y(M_[0]);
+
+    y_vi_
+        = ChainableStack::instance().memalloc_.alloc_array<vari*>(var_y.size());
+    for (size_t m = 0; m < var_y.size(); ++m) {
+      y_vi_[m] = new vari(val_y[m], false);
+      var_y[m] = y_vi_[m];
```

Comment:
[question]

Is there no way to get away with only a single container here because `y_vi_` itself is used as a container as is `var_y` in the future?